### PR TITLE
Compatibility let syntax

### PR DESCRIPTION
**Pull request by ghost**
Implements #1891

### CODE REVIEW

**Comment by diml** [comment]
`and [@toto] ...` is not a valid syntax in 4.02 :(

We need to find another way to retrofit the syntax...

**Comment by emillon** [comment]
In the meantime do you think it could be worth updating the existing preprocessor syntax first, then figuring out how to make it generic? I'm not familiar enough with the processor to understand if it's a 2 line change or something more complicated.

**Comment by diml** [comment]
Actually, I found a way to make it work. I store custom operators in a hash table indexed by the starting position of the `let` and `and` keywords.

**Comment by diml** [comment]
@dra27 @nojb, does this command line looks correct to you:

```
"C:/PROGRA~1/OCaml\bin\ocamldep.opt.exe" -modules -pp "./boot-pp.exe -dump-ast" src\action.ml > boot-depends.txt
```

It fails in AppVeyor and I don't really understand why. The only new thing in this PR is the `-pp ...` flag.

**Comment by nojb** [comment]
Maybe a problem locating `ocamlrun`?

**Comment by diml** [comment]
Possibly. I tried passing `-custom`, let see it that helps

**Comment by diml** [comment]
Nope, that didn't help

**Comment by nojb** [comment]
Actually the error appears at the very beginning of the run: https://ci.appveyor.com/project/diml/dune/builds/22804426#L13.

**Comment by diml** [comment]
Indeed. I just checked previous successful builds and the same error is present. Maybe the error message is not directly related.

**Comment by nojb** [comment]
May be a silly suggestion, but did you try replacing `./boot_pp.exe` by `.\boot_pp.exe`?

**Comment by diml** [comment]
That seems like a good thing to try indeed. I'm wondering if quoting the program name is not wrong as well

**Comment by nojb** [comment]
Did you try to reproduce locally?

**Comment by nojb** [comment]
I can reproduce locally. I am investigating.

**Comment by diml** [comment]
@nojb not quoting the program name and replacing `./` by `.\` seems to fix the issue in AppVeyor

**Comment by diml** [comment]
Ok, CI's now passing. This is ready for review. The diff is big, but most of it is just upgrading to the new syntax.

File: doc/dune-files.rst
Comment (emillon) [inline doc/dune-files.rst]: How about `compat-syntax`? or `future-syntax`? (that's similar to what python uses)

File: test/blackbox-tests/test-cases/github25/run.t
Comment (emillon) [inline test/blackbox-tests/test-cases/github25/run.t]: It's a bit strange that this changes the output. Do you know why that is the case?

File: doc/dune-files.rst
Comment (emillon) [inline doc/dune-files.rst]:
```suggestion
``no_preprocessing`` when using a recent version of the
```

(should this be "the latest"?)

File: doc/dune-files.rst
Comment (diml) [reply doc/dune-files.rst]: I like `future-syntax`, let's go for that

Comment (diml) [reply doc/dune-files.rst]: I'm using `future_syntax` to be consistent with the other cases

File: doc/dune-files.rst
Comment (diml) [reply doc/dune-files.rst]: note that it can be several versions of the compiler

Comment (diml) [reply doc/dune-files.rst]: I change this sentence

File: test/blackbox-tests/test-cases/github25/run.t
Comment (diml) [reply test/blackbox-tests/test-cases/github25/run.t]: Arguments are collected into a set and the compare function of this set compares ids. From cmdliner_info.ml:

```ocaml
module Arg = struct
  type t = arg
  let compare a0 a1 = (compare : int -> int -> int) a0.id a1.id
end

module Args = Set.Make (Arg)
```

This PR changes a bit the shape of the command line description, from `const f $ arg1 $ arg2 $ arg3` to `const f $ ((and+ arg1 arg2) arg3)`. Given that `arg{1,2,3}` are not evaluated in the same order anymore, the identifiers are not generated in the same order as well and this changes the order of arguments in the set.

Comment (diml) [reply test/blackbox-tests/test-cases/github25/run.t]: There is not much we can do for now without changing cmdliner first.

Comment (emillon) [reply test/blackbox-tests/test-cases/github25/run.t]: Thanks for the explanation. Is this guaranteed to be evaluated in the same order as the native version in the compiler? Otherwise, would it make sense to fix the order of evaluation by emitting `let v1 = arg1 in let v2 = arg2 in and+ v1 v2`?

Comment (diml) [reply test/blackbox-tests/test-cases/github25/run.t]: We try to avoid these kinds of differences in behavior between native and bytecode, however there is no strong guarantee that they match. Regarding the rewriting, we should match what 4.08 does. After playing with the toplevel, it does seem that  we should generate these bindings:

```ocaml
# let (let+) x f = f x;;
val ( let+ ) : 'a -> ('a -> 'b) -> 'b = <fun>
# let (and+) a b = a, b;;
val ( and+ ) : 'a -> 'b -> 'a * 'b = <fun>
# let+ () = print_endline "a" and+ () = print_endline "b" and+ () = print_endline "c" in ();;
a
b
c
- : unit = ()
# (let+) 1 ignore;;
- : unit = ()
# (let+) ((and+) ((and+) (print_endline "a") (print_endline "b")) (print_endline "c")) (fun (((), ()), ()) -> ());;
c
b
a
```

**Comment by diml** [comment]
I updated the bootstrap to use ocamlopt if possible to build the bootstrap preprocessor. This cuts the bootstrap time by 3s.

File: doc/dune-files.rst
Comment (emillon) [inline doc/dune-files.rst]:
```suggestion
  OCaml syntaxes to older compilers. See :ref:`Future syntax
```

File: doc/dune-files.rst
Comment (emillon) [inline doc/dune-files.rst]:
```suggestion
backports some of the newer syntax elements. This allows you to use some of
```

**Comment by emillon** [comment]
Cool!

I was wondering if there are some versioning aspects that we should consider (ie should there be a relation between the dune project version and the syntax elements that are polyfilled). Let's say we add a polyfill for `let exception E in` in dune 1.9. How do dune users express that they need this feature?

**Comment by diml** [comment]
Versioning this seems good, we could simply tie it to the dune language version. However, I'm still unsure how much we can provide in dune: the only available hook we have to extend the syntax is post-processing the token stream, which is not much. It works well for `let+`, but for  instance for `let exception E in` we need to recognize sequences of tokens that match `let exception <uid> [of <type_expr>] in`, which is already a lot more complex.

The right way to provide this feature is to have an OCaml parser with the grammar of the latest version of OCaml but the AST of the current version being used. This is a much bigger project and probably not something we want to maintain inside Dune.

**Comment by emillon** [comment]
Yes, I agree that we shouldn't polyfill every bit of syntax. And versioning can probably be dealt with when we upgrade the preprocessor.

**Comment by diml** [comment]
Indeed. @emillon, I'm now just waiting for you to Approve the PR before merging :)

**Review by emillon — APPROVED** [review]
Ah, didn't realize this :)

**Pull request merged**
